chunk.cpp

```diff
@@ -30,46 +32,39 @@ void Chunk::load(const NBT &nbt) {
     this->sections[i] = NULL;
   highest = 0;
 
-  auto level = nbt.at("Level");
+  int version = nbt.at("DataVersion")->toInt();
+  const Tag * level = nbt.at("Level");
   chunkX = level->at("xPos")->toInt();
   chunkZ = level->at("zPos")->toInt();
 
+  // load Biome per column
   auto biomes = level->at("Biomes");
-  memcpy(this->biomes, biomes->toIntArray(), 4*biomes->length());
+  if (version >= 1519) {
+    memcpy(this->biomes, biomes->toIntArray(), sizeof(int)*biomes->length());
+  } else {
+    // convert quint8 to quint32
+    auto rawBiomes = biomes->toByteArray();
+    for (int i=0; i<256; i++)
+      this->biomes[i] = rawBiomes[i];
+  }
+
+  // load available Sections
   auto sections = level->at("Sections");
   int numSections = sections->length();
-  for (int i = 0; i < numSections; i++) {
-    auto section = sections->at(i);
-    auto cs = new ChunkSection();
-    // decode Palette to be able to map BlockStates
-    auto rawPalette = section->at("Palette");
-    cs->paletteLength = rawPalette->length();
-    cs->palette = new BlockData[cs->paletteLength];
-    for (int j = 0; j < rawPalette->length(); j++) {
-      cs->palette[j].name = rawPalette->at(j)->at("Name")->toString();
-      if (rawPalette->at(j)->has("Properties"))
-        cs->palette[j].properties = rawPalette->at(j)->at("Properties")->getData().toMap();
-    }
-    // map BlockStates to BlockData
-    // todo: bit fidling looks very complicated -> find easier code
```

Comment:
I tinkered around with the bit fiddling to remove the need to call `std::reverse` in my latest commit to #127. 

The data in BlockStates is a bit-reversed array stored in little endian format. So for a 4 block array, where each block is six bits, looks like this:

```
BBAAAAAA CCCCBBBB DDDDDDCC
```

If each block had a value of `5 (000101)`, then this example array looks like this:

```
BBAAAAAA CCCCBBBB DDDDDDCC
00101000 10001010 10100010
```

The algorithm here will byte-swap the array, so that the array looks to be in order; from there I can grab the bits by passing in the index and the number of bits I want to grab. In the example below, to grab block C, I would call `getBits(data, 6, 6)` on the byte-swapped array, which looks like this:

```
DDDDDDCC CCCCBBBB BBAAAAAA
```

`getBits` concatenates enough bytes together (1) so that I'm guaranteed the whole data I want is in the same word. It then shifts this combined word into place (2) and masks it (3) to obtain the data I want to return (4):

```
(1) DDDDDDCCCCCCBBBBBBAAAAAA <- The concatenated word
(2) 000000000000DDDDDDCCCCCC <- Shift the word so that the data is in the lowest bit
(3) 000000000000000000111111 <- The mask produced by ((1 << n) - 1)
(4) 000000000000000000CCCCCC <- Result of (2) & (3)
```

I hope this information can help lead to a more efficient algorithm. The commit I recently made changes `getBits` to operate on the little endian array so the byte-swap isn't necessary. There is definitely room to clear up this algorithm.

Reply:
Mojang is using Little Endian, how strange is that?
I think it is a good idea to take a look to the code they use to decode that part. Mostly they have time to develop nice and efficient algorithms. But this is also something for later this year...

Just as comment for later:
A section can only hold 16*16*16 blocks, so at maximum 4096 different blocks can reside in a palette. This leads to maximum of 12 bits that has to be supported. 12 bits can spread at maximum over 3 bytes.
Typical values are probably 3..6 bits.

Reply:
I'm curious to find out how Mojang parses the data. I would agree that using their algorithm (if possible) is the best route going forward. These patches to Minutor makes it one of the first mapping tools to support block-level information about 1.13 worlds that I know of.

Also, if each block was unique, the palette would have 4096 entries, which requires 13 bits (a one followed by 12 zeros). This is still spread at maximum over three bytes.